File: src/lib/okx-client.ts
import { createHmac } from 'node:crypto';
import { AsyncLocalStorage } from 'node:async_hooks';
import { readSnapshot, writeSnapshot } from './snapshot';
export const pause = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
let queue: Promise<unknown> = Promise.resolve();
let nextAt = 0;
let loaded = false;
type Allowance={remaining:number;parent?:Allowance};
const allowance=new AsyncLocalStorage<Allowance>();
export function withRequestAllowance<T>(limit:number,action:()=>T):T{return allowance.run({remaining:limit,parent:allowance.getStore()},action);}
const usage={day:'',daily:0,round:0,startedAt:0,completedAt:0,nextAt:0,lastError:null as string|null};
const positive=(value:string|undefined,fallback:number)=>Number.isFinite(Number(value))&&Number(value)>0?Math.floor(Number(value)):fallback;
export function collectionStatus(){return {...usage,dailyLimit:positive(process.env.OKX_DAILY_REQUEST_LIMIT,8000),roundLimit:positive(process.env.OKX_ROUND_REQUEST_LIMIT,70),intervalMs:300000};}
export function startCollection(){usage.round=0;usage.startedAt=Date.now();usage.lastError=null;usage.nextAt=0;}
export function endCollection(error:string|null=null){usage.completedAt=Date.now();usage.nextAt=Date.now()+300000;usage.lastError=error;}
function chargeRequest(){
  if(!loaded){if(process.env.NODE_ENV!=='test')Object.assign(usage,readSnapshot('data/okx-usage.json')??{}, {round:usage.round,startedAt:usage.startedAt});loaded=true;}
  const day=new Date().toISOString().slice(0,10);if(usage.day!==day){usage.day=day;usage.daily=0;}
  const limits=collectionStatus();
  const budgets:Allowance[]=[];for(let b=allowance.getStore();b;b=b.parent)budgets.push(b);
  if(budgets.some(b=>b.remaining<=0))throw new Error('OKX network request allowance exhausted');
  if(usage.daily>=limits.dailyLimit||usage.round>=limits.roundLimit)throw new Error('OKX local request budget exhausted');
  for(const budget of budgets)budget.remaining--;
  usage.daily++;usage.round++;
  if(process.env.NODE_ENV!=='test')writeSnapshot('data/okx-usage.json',usage);
}

// All workers share one rate limiter. Never send these headers to another host.
export function okxGet(endpoint: string, params: Record<string, string>): Promise<any> {
  return okxRequest(endpoint + '?' + new URLSearchParams(params), 'GET');
}
export function okxPost(endpoint: string, data: unknown): Promise<any> {
  return okxRequest(endpoint, 'POST', JSON.stringify(data));
}
function okxRequest(path: string, method: 'GET' | 'POST', body = ''): Promise<any> {
  const task = async () => {
    const key = process.env.OKX_API_KEY, secret = process.env.OKX_SECRET_KEY, passphrase = process.env.OKX_PASSPHRASE;
    if (!key || !secret || !passphrase) throw new Error('OKX credentials not configured');
    if (!path.startsWith('/api/v6/dex/')) throw new Error('Invalid OKX endpoint');
    for (let attempt = 0; attempt < 4; attempt++) {
      await pause(Math.max(0, nextAt - Date.now()));
      chargeRequest();
      const timestamp = new Date().toISOString();
      const response = await fetch('https://web3.okx.com' + path, {
        method, ...(body ? {body} : {}),
        headers: { 'Content-Type': 'application/json', 'OK-ACCESS-KEY': key, 'OK-ACCESS-PASSPHRASE': passphrase,
          'OK-ACCESS-TIMESTAMP': timestamp,
          'OK-ACCESS-SIGN': createHmac('sha256', secret).update(timestamp + method + path + body).digest('base64') },
        signal: AbortSignal.timeout(15000),
      });
      nextAt = Date.now() + 400;
      const json = await response.json().catch(() => ({}));
      if ((response.status === 429 || json.code === '50011') && attempt < 3) {
        nextAt = Date.now() + 1000 * 2 ** attempt;
        continue;
      }
      if (!response.ok || json.code !== '0') {
        // Upstream diagnostics may contain echoed request fields: expose codes only.
        throw new Error(`OKX HTTP ${response.status} · code ${String(json.code ?? 'unknown').replace(/[^\w-]/g, '').slice(0, 30)}`);
      }
      return json.data;
    }
    throw new Error('OKX rate limit');
  };
  const result = queue.then(task);
  queue = result.catch(() => {});
  return result;
}

File: server-py/app/okx_client.py
"""OKX DEX API client: HMAC signing, local quota ledger, throttling and
backoff — a port of the Node okx-client. The round budget guards scan-type
bursts; price-info batches may skip it (skip_round) but never the daily cap."""
import asyncio
import base64
import hashlib
import hmac
import json
import os
import time
from contextlib import contextmanager
from datetime import datetime, timezone

import httpx

BASE = "https://web3.okx.com"
USAGE_FILE = "data/okx-usage.json"


def _positive(env: str | None, fallback: int) -> int:
    try:
        v = int(os.environ.get(env, "") or 0)
        return v if v > 0 else fallback
    except ValueError:
        return fallback


class Quota:
    def __init__(self):
        self.day = ""
        self.daily = 0
        self.round = 0
        self.last_persist = 0.0
        self._loaded = False
        self._persist_task: asyncio.Task | None = None

    def limits(self):
        return {
            "daily": _positive("OKX_DAILY_REQUEST_LIMIT", 8000),
            "round": _positive("OKX_ROUND_REQUEST_LIMIT", 70),
        }

    def rollover(self):
        today = datetime.now(timezone.utc).strftime("%Y-%m-%d")
        if self.day != today:
            self.day = today
            self.daily = 0

    def load(self):
        if self._loaded:
            return
        self._loaded = True
        if os.environ.get("NODE_ENV") == "test" or not os.path.exists(USAGE_FILE):
            return
        try:
            saved = json.load(open(USAGE_FILE))
            self.day = saved.get("day", "")
            self.daily = saved.get("daily", 0)
        except Exception:
            pass

    def persist_soon(self):
        if os.environ.get("NODE_ENV") == "test":
            return

        def write():
            try:
                os.makedirs(os.path.dirname(USAGE_FILE), exist_ok=True)
                tmp = USAGE_FILE + ".tmp"
                with open(tmp, "w") as f:
                    json.dump({"day": self.day, "daily": self.daily}, f)
                os.replace(tmp, USAGE_FILE)
            except Exception:
                pass

        now = time.time()
        if now - self.last_persist >= 30:
            self.last_persist = now
            write()
            return
        if not self._persist_task or self._persist_task.done():
            async def later():
                await asyncio.sleep(30)
                self.last_persist = time.time()
                write()
            try:
                self._persist_task = asyncio.create_task(later())
            except RuntimeError:
                pass


USAGE = Quota()

# Allowance contexts mirror the Node AsyncLocalStorage budgets.
_allowance_set: set[asyncio.Task] = set()


@contextmanager
def request_allowance(limit: int):
    task = asyncio.current_task()
    token = (task, limit)
    _allowance_set.add(token)
    try:
        yield
    finally:
        _allowance_set.discard(token)


def _remaining_allowance() -> int | None:
    task = asyncio.current_task()
    vals = [limit for (t, limit) in _allowance_set if t is task]
    return min(vals) if vals else None


class QuotaExceeded(Exception):
    pass


def charge(skip_round: bool = False):
    USAGE.load()
    USAGE.rollover()
    limits = USAGE.limits()
    remaining = _remaining_allowance()
    if remaining is not None and remaining <= 0:
        raise QuotaExceeded("OKX network request allowance exhausted")
    if USAGE.daily >= limits["daily"] or (not skip_round and USAGE.round >= limits["round"]):
        raise QuotaExceeded("OKX local request budget exhausted")
    USAGE.daily += 1
    if not skip_round:
        USAGE.round += 1
    USAGE.persist_soon()


def reset_round():
    USAGE.round = 0


_serial_lock = asyncio.Lock()
_throttle = {"next_at": 0.0}


async def okx_get(endpoint: str, params: dict[str, str]):
    path = endpoint + "?" + "&".join(f"{k}={v}" for k, v in params.items())
    return await _request(path, "GET", "")


async def okx_post(endpoint: str, data, opts: dict | None = None):
    opts = opts or {}
    body = json.dumps(data, separators=(",", ":"))
    return await _request(endpoint, "POST", body, opts)


async def _request(path: str, method: str, body: str, opts: dict | None = None):
    opts = opts or {}
    key = os.environ.get("OKX_API_KEY")
    secret = os.environ.get("OKX_SECRET_KEY")
    passphrase = os.environ.get("OKX_PASSPHRASE")
    if not (key and secret and passphrase):
        raise RuntimeError("OKX credentials not configured")
    if not path.startswith("/api/v6/dex/"):
        raise RuntimeError("Invalid OKX endpoint")

    skip = bool(opts.get("skip_round"))
    urgent = bool(opts.get("urgent"))

    async def attempt():
        await asyncio.sleep(max(0.0, _throttle["next_at"] - time.time()))
        charge(skip)
        timestamp = datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
        message = (timestamp + method + path.split("?")[0] + (body if method == "POST" else "")).encode()
        if method == "GET" and path.split("?", 1)[1:]:  # signed query string included for GET
            message = (timestamp + method + path).encode()
        signature = base64.b64encode(hmac.new(secret.encode(), message, hashlib.sha256).digest()).decode()
        headers = {
            "Content-Type": "application/json",
            "OK-ACCESS-KEY": key,
            "OK-ACCESS-PASSPHRASE": passphrase,
            "OK-ACCESS-TIMESTAMP": timestamp,
            "OK-ACCESS-SIGN": signature,
            "User-Agent": "curl/8.7.1",
        }
        async with httpx.AsyncClient(timeout=15.0) as client:
            if method == "GET":
                resp = await client.get(BASE + path, headers=headers)
            else:
                resp = await client.post(BASE + path, content=body, headers=headers)
        _throttle["next_at"] = time.time() + 0.4
        payload = {}
        try:
            payload = resp.json()
        except Exception:
            pass
        if resp.status_code == 429 or payload.get("code") == "50011":
            raise Retryable()
        if resp.status_code >= 400 or payload.get("code") not in (None, "0"):
            detail = str(payload.get("code", resp.status_code))[:30]
            raise RuntimeError(f"OKX HTTP {resp.status_code} · {detail}")
        return payload.get("data")

    class Retryable(Exception):
        pass

    async def guarded():
        # The serial lock preserves the Node promise-queue semantics except
        # for urgent batched price calls, which bypass it.
        if urgent:
            return await attempt()
        async with _serial_lock:
            return await attempt()

    backoff = 1.0
    for i in range(4):
        try:
            return await guarded()
        except Retryable:
            if i == 3:
                raise RuntimeError("OKX rate limit")
            _throttle["next_at"] = time.time() + backoff
            backoff *= 2
            await asyncio.sleep(backoff)
        except QuotaExceeded:
            raise
